File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.cpp
#include <bits/stdc++.h>

using namespace std;


int cnt[51][51];

vector<string> split(string str) {
    vector<string> ret;
    stringstream ss(str);
    string s;
    while(getline(ss, s, ' ')) {
        ret.push_back(s);
    }
    return ret;
}

int findIdx(vector<string> friends, string name) {
    for(int i = 0; i < friends.size(); i++) {
        if(friends[i] == name) {
            return i;
        }
    }
}

int solution(vector<string> friends, vector<string> gifts) {
    int answer = 0;
    
    for(int i = 0; i < gifts.size(); i++) {
   		vector<string> name = split(gifts[i]); 
       	int idx1 = findIdx(friends, name[0]); 
        int idx2 = findIdx(friends, name[1]);
        cnt[idx1][idx2]++;
    }
    
    int sum[51][2] = {0, };
    for(int i = 0; i < friends.size(); i++) {
        for(int j = 0; j < friends.size(); j++) {
            sum[i][0] += cnt[i][j];
            sum[i][1] += cnt[j][i];
        }
    }
    
   	int res[51] = {0, };
    for(int i = 0; i < friends.size(); i++) {
        for(int j = 0; j < friends.size(); j++) {
            if(i == j) continue;
            if(cnt[i][j] > cnt[j][i]) res[i]++;
            else if((!cnt[i][j] && !cnt[j][i]) || cnt[i][j] == cnt[j][i]) {
                int score1 = sum[i][0] - sum[i][1];
                int score2 = sum[j][0] - sum[j][1];
                if(score1 > score2)
                  	res[i]++;
            }
        }
    }
    
    answer = *max_element(res, res + friends.size());
    
    return answer;
}

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;
import java.util.stream.*;

class Solution {
   	static Map<String, Integer> map = new HashMap<>(); 
    static int[][] arr = new int[51][51];
    static int[] point = new int[51];
    static int[] cnt = new int[51];
    static int n;
    
    static void solve() {
       	for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (i == j) continue;
                if(arr[i][j] > arr[j][i]) {
                   	cnt[i]++; 
                } else if(arr[i][j] == arr[j][i] && point[i] > point[j]) {
                  	cnt[i]++; 
                }
            }
        }
    }
    
    public int solution(String[] friends, String[] gifts) {
        int answer = 0;
        n = friends.length;
        
       	for (int i = 0; i < n; i++) {
            map.put(friends[i], i);
        } 
        
        for (int i = 0; i < gifts.length; i++) {
            String[] tmp = gifts[i].split(" ");
            int idx1 = map.get(tmp[0]);
            int idx2 = map.get(tmp[1]);
           	
            arr[idx1][idx2]++;
        }
        
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
              	point[i] += arr[i][j];
                point[i] -= arr[j][i];
            }
            
        }
        solve();
        
        answer = Arrays.stream(cnt).max().getAsInt();
        
        return answer;
    }
}